ndk-build/src/manifest.rs

```diff
@@ -1,258 +1,345 @@
-            android:label="{package_label}"
-            android:debuggable="{debuggable}"
-            {icon}
-            {fullscreen}>
-            {application_metadatas}
-        <activity
-                android:name="android.app.NativeActivity"
-                android:label="{package_label}"
-                android:screenOrientation="{orientation}"
-                android:launchMode="{launch_mode}"
-                android:configChanges="orientation|keyboardHidden|screenSize">
-            <meta-data android:name="android.app.lib_name" android:value="{target_name}" />
-            {activity_metadatas}
-            <intent-filter>
-                <action android:name="android.intent.action.MAIN" />
-                <category android:name="android.intent.category.LAUNCHER" />
-            </intent-filter>
-            {intent_filters}
-        </activity>
-    </application>
-</manifest>"#,
-            package_name = &self.package_name,
-            package_label = &self.package_label,
-            version_name = &self.version_name,
-            version_code = self.version_code,
-            split = split,
-            target_sdk_version = self.target_sdk_version,
-            min_sdk_version = self.min_sdk_version,
-            opengles_version = opengles_version,
-            target_name = &self.target_name,
-            icon = icon,
-            fullscreen = fullscreen,
-            orientation = orientation,
-            launch_mode = launch_mode,
-            application_metadatas = application_metadatas.join("\n"),
-            activity_metadatas = activity_metadatas.join("\n"),
-            debuggable = self.debuggable,
-            features = features.join("\n"),
-            permissions = permissions.join("\n"),
-            intent_filters = intent_filters.join("\n"),
-        )
+fn serialize_features<S>(features: &Option<Vec<Feature>>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    // Check if features contains an `opengles_version` if not, add a default feature.
+    match features {
+        Some(features) => {
+            if features.iter().all(|f| f.opengles_version.is_none()) {
+                let mut f = features.clone();
+                f.push(Feature {
+                    name: None,
+                    required: Some(true),
+                    opengles_version: Some((3, 1)),
+                });
+                return serializer.serialize_some(&f);
+            }
+            serializer.serialize_some(features)
+        }
+        None => serializer.serialize_none(),
     }
+}
 
+impl AndroidManifest {
     pub fn write_to(&self, dir: &Path) -> Result<(), NdkError> {
         let mut file = File::create(dir.join("AndroidManifest.xml"))?;
-        writeln!(file, "{}", self.to_string())?;
+        let serialized = quick_xml::se::to_string(&self).unwrap();
+        writeln!(file, r#"<?xml version="1.0" encoding="utf-8"?>"#)?;
+        writeln!(file, "{}", serialized)?;
```

Comment:
You shouldn't need to worry about the XML declaration and can instead serialize to the writer directly:
```suggestion
        quick_xml::se::to_writer(file, &self).unwrap();
```

Should we add the error type to `NdkError` so that it can be `?`'d instead of unwrapping?

Reply:
Done!

Reply:
So we pushed a similar change in our own codebase where we were writing 60+MB JSON files and guess what, everything slowed down to a halt!

The serializer performs a write for every little thing that it emits, simply because that is how Serde is structured. Since I'm not a huge fan of serializing anything to a string followed by dumping that to a file the best of both worlds is probably a `BufWriter`, that abstracts away the buffering logic you'd get with `to_string` but with a sensible size, performing infrequent writes:

```rust
let file = File::create(dir.join("AndroidManifest.xml"))?;
let w = std::io::BufWriter::new(file);
quick_xml::se::to_writer(w, &self)?;
```

Reply:
> completely looked over it

NP: It was my suggestion and I didn't realize this either :). Rusts file writer is just a plain writer that does exactly what it says on the tin - execute the `write()` syscall every time it is called without any hidden batching. Same for Serde, minimizing its own runtime overhead and putting the burden of caching on the writer-backend - which would be suboptimal if writing directly to memory anyway. That is exactly what happens when serializing to a string, that's backed by a `Vec<u8>`! Pretty neat and explicit if you ask me.

Even if these files are small, that count still adds up if a write is invoked for every primitive (`<`/`>`, indentation and spaces, key and value strings, `=`, etc).